packages/assets-controllers/src/assetsUtil.ts

```diff
@@ -228,3 +236,65 @@ export function addUrlProtocolPrefix(urlString: string): string {
 export function ethersBigNumberToBN(bigNumber: BigNumber): BN {
   return new BN(stripHexPrefix(bigNumber.toHexString()), 'hex');
 }
+
+/**
+ * Partitions a list of values into groups that are at most `batchSize` in
+ * length.
+ *
+ * @param values - The list of values.
+ * @param args - The remaining arguments.
+ * @param args.batchSize - The desired maximum number of values per batch.
+ * @returns The list of batches.
+ */
+export function divideIntoBatches<Value>(
+  values: Value[],
+  { batchSize }: { batchSize: number },
+): Value[][] {
+  const batches = [];
+  for (let i = 0; i < values.length; i += batchSize) {
+    batches.push(values.slice(i, i + batchSize));
+  }
+  return batches;
+}
+
+/**
+ * Constructs a data structure from processing batches of the given values
+ * sequentially.
+ *
+ * @param args - The arguments to this function.
+ * @param args.values - A list of values to iterate over.
+ * @param args.batchSize - The maximum number of values in each batch.
+ * @param args.eachBatch - A function to call for each batch. This function is
+ * similar to the function that `Array.prototype.reduce` takes, in that
+ * it receives the data structure that is being built along with the batch per
+ * iteration through the list of batches, and should return an updated version
+ * of the data structure.
+ * @param args.initialResult - The initial value of the final data structure,
+ * i.e., the value that will be fed into the first call of `eachBatch`.
+ * @returns The final data structure.
+ */
+export async function reduceInBatchesSerially<Value, Result>({
+  values,
+  batchSize,
+  eachBatch,
+  initialResult,
+}: {
+  values: Value[];
+  batchSize: number;
+  eachBatch: (
+    workingResult: PartialObject<Result>,
```

Comment:
Nit: I think we could use `Partial` here rather than `PartialObject`. Maybe we don't need this `PartialObject` type?

It seems that this type is something like "If object, then Partial, else return the type unchanged". But in this case, we can require that `Result` is an object.

Perhaps you meant for this to be generic over the result type, but I'm not sure that's worth the added type complexity, as the `PartialObject` wouldn't work effectively for any arbitrary type of collection anyway.

Reply:
Yes, I meant for this utility to be generic so that it can be used in other cases where we might want to batch requests. So you should be able to use it to build not only an object but also an array. That's why `PartialObject` returns the type unchanged if it's not an object type, because arrays don't get partially filled until they're completed, they just get extended. But I suppose we don't need to make this generic right now.

Reply:
Updated to be less generic and use `Partial` here: [07c4ff4](https://github.com/MetaMask/core/pull/3650/commits/07c4ff49d6250a488227323663331a335ce0c5a2)